### ISSUE TITLE
Fix S3900 FP: Parameter re-assignment through coalescing assignment and tuple deconstruction

### ISSUE THREAD

**Issue by zsolt-kolbay-sonarsource** [SonarSource/sonar-dotnet#7091]
### Description

Currently, the S3900 rule does not raise an issue if the parameter of the method is assigned a new value before it's dereferenced:

```cs
public void Method(object o)
{
    o = Unknown();
    o.ToString(); // Compliant - o no longer has the value which was passed to the method
}

private object Unkown() => null;
```

This currently works for simple assignments, but not for coalescing assignments and tuple deconstructions.

- [ ] Coalescing assignment:
```cs
public void CoalescingAssignment(object o)
{
    o ??= Unknown();
    o.ToString(); // Noncompliant - FP
}

- [ ] Tuple deconstruction:
```cs
public void TupleDeconstruction(object o)
{
    (o, _) = (Unknown(), Unknown());
    o.ToString(); // Noncompliant - FP
}

private object Unkown() => null;
```

### Related information

* C#/VB.NET Plugins version: 8.56
* Visual Studio version: Visual Studio 2022 17.5
* Operating System: Windows 10

**Comment by martin-strecker-sonarsource** [comment]
Closed as tracked by https://sonarsource.atlassian.net/browse/NET-1200 now.
